src/opnmidi_private.hpp

```diff
@@ -604,37 +741,115 @@ class OPNMIDIplay
     // Additional information about OPN channels
     struct OpnChannel
     {
-        // For collisions
         struct Location
         {
             uint16_t    MidCh;
             uint8_t     note;
-            bool operator==(const Location &b) const
-            {
-                return MidCh == b.MidCh && note == b.note;
-            }
-            bool operator< (const Location &b) const
-            {
-                return MidCh < b.MidCh || (MidCh == b.MidCh && note < b.note);
-            }
+            bool operator==(const Location &l) const
+                { return MidCh == l.MidCh && note == l.note; }
+            bool operator!=(const Location &l) const
+                { return !operator==(l); }
             char ____padding[1];
         };
         struct LocationData
         {
+            LocationData *prev, *next;
+            Location loc;
             bool sustained;
             char ____padding[1];
             MIDIchannel::NoteInfo::Phys ins;  // a copy of that in phys[]
             char ____padding2[4];
             int64_t kon_time_until_neglible;
             int64_t vibdelay;
         };
-        typedef std::map<Location, LocationData> users_t;
-        users_t users;
 
         // If the channel is keyoff'd
         int64_t koff_time_until_neglible;
+
+        enum { users_max = 128 };
+        LocationData *users_first, *users_free_cells;
+        LocationData users_cells[users_max];
+        unsigned users_size;
+
+        bool users_empty() const
+            { return !users_first; }
+        LocationData *users_find(Location loc)
+        {
+            LocationData *user = NULL;
+            for(LocationData *curr = users_first; !user && curr; curr = curr->next)
+                if(curr->loc == loc)
+                    user = curr;
+            return user;
+        }
+        LocationData *users_allocate()
+        {
+            // remove free cells front
+            LocationData *user = users_free_cells;
+            if(!user)
+                return NULL;
+            users_free_cells = user->next;
+            users_free_cells->prev = NULL;
```

Comment:
On Android just now I have caugh SIGSEGV crash because `users_free_cells` is NULL :laughing:

Reply:
And when I have added a condition to `users_free_cells` is there are null, I have got another crash:
![default](https://user-images.githubusercontent.com/6751442/39089934-7c075e0a-45da-11e8-9ae4-a84cd227c6c0.png)

Reply:
`users_free_cells` must definitely be checked for `NULL` before accessing `prev`.
This stack trace comes from the android phone?
Do you have the midi file to produce this error condition?

Reply:
@Wohlstand update: did you notice note=0x9b (decimal 155), which is over midi limit of 127, going out of the array bounds?

Reply:
@jpcima , Yes, that came from a phone, I ran the debug on my phone.
Note=0x9b looking weird... yeah, gotta to re-check again, then I'll send the MIDI file itself when crash will happen again.

Reply:
It's possible technically that MIDI files contain note events with values on 8bits... it's not standard.
You can check this on your program side, or change ADLMIDI to filter the bad events.
None of the current rt APIs taking uint8s perform any check.

Reply:
Ha-ha!
```cpp
if(!d)
    d->xxx();
```
Did you meant:
```cpp
if(d == NULL)
    d->xxx();
```
? :laughing: 
![default](https://user-images.githubusercontent.com/6751442/39093126-02fd3d42-4623-11e8-90fc-33619ad3b702.png)

Reply:
I have caught this crash from VLC when I ran it in debug mode with puting of symlink to my debug build of plugin.

Reply:
I see. It's oversight that `d` was used in the middle of body, I thought it wasn't.
This part is a pitch bend handler.
My rationale for `!d || !d->sustained` was to think that: if I don't have an information on this note, I will not assume it is sustained or not, and I will just let pitch bend apply.

In the same rationale, I could maybe say: (allow vibrato to apply)
`if(Ch[MidCh].vibrato && (!d || d->vibdelay >= Ch[MidCh].vibdelay))`

Reply:
> Yes, and I think, I really must to make a unit-test now, quickly, to debug that perfect map

Sorry for mistakes. It's frustrating to fix the silly problems.
If it were me, I would go for boost data structures and not reinvent this linked list code. If I wanted to spend more time I would write it in C++ generic data structure fashion. It would be much simpler for a unit test.